### DISCUSSION TITLE
Fehlende Batterie in der Live-Übersicht

### DISCUSSION THREAD

**Discussion by sentinel1111** [hoylabs/OpenDTU-OnBattery#1151] [General]
Ich habe mir eine neue Version für den Esp32S3 compiliert.
Nach der OTA Firmware-Aktualisierung fehlt mir in der Live Übersicht der Pylontech und auch in den Einstellungen,beim Stromzähler, fehlt der Http(s)/Json Eingabefeld.
In der Console sind die Daten von der Batterie und von Stromzähler zu sehen,und auch ganz oben, in den Fenster bei Live-View sind
Ladezustand und Leistung/Netz  Daten vorhanden.
Weiß jemand was ich falsch mache ? Cache geleert und verschiedene Browser ausprobiert.

[log_opendtu.txt](https://github.com/user-attachments/files/16487742/log_opendtu.txt)

**Comment by schlimmchen** [comment]
https://www.opendtu.solar/firmware/compile_webapp/

Ich denke ich werde die alten webapp Artefakte rausschmeißen aus dem repo, sodass man auf die finger kriegt beim Kompilieren, weil die Dateien fehlen. Das hat ja keinem Sinn das jede Woche zwei mal erzählen zu müssen, dass die webapp sich nicht selbst kompiliert 🙊

**Reply by spcqike** [reply]
> dass die webapp sich nicht selbst kompiliert

@schlimmchen kann man nicht ein pre-Skript dafür laufen lassen? welches den Hash aller Datein in ``webapp`` ermittelt und prüft, ob sich einer geändert hat?

```
import os
import hashlib
import pickle
import subprocess

# Funktion zur Berechnung des Hash-Werts einer Datei
def calculate_hash(file_path):
    hasher = hashlib.md5()
    with open(file_path, 'rb') as f:
        buf = f.read()
        hasher.update(buf)
    return hasher.hexdigest()

# Funktion zur Überprüfung der Dateien in einem Ordner
def check_files(directory, hash_file):
    file_hashes = {}

    # Alle Dateien im Ordner und Unterordner durchsuchen
    for root, dirs, files in os.walk(directory):
        print(f"Überprüfe Dateien in {root}")
        for file in files:
            file_path = os.path.join(root, file)
            file_hashes[file_path] = calculate_hash(file_path)
    
    # Laden der vorherigen Hash-Werte
    if os.path.exists(hash_file):
        with open(hash_file, 'rb') as f:
            old_file_hashes = pickle.load(f)
    else:
        old_file_hashes = {}
    
    # Vergleich der aktuellen Hash-Werte mit den vorherigen
    for file_path, file_hash in file_hashes.items():
        if file_path not in old_file_hashes or old_file_hashes[file_path] != file_hash:
            execute_command()
            break
    
    # Speichern der aktuellen Hash-Werte
    with open(hash_file, 'wb') as f:
        pickle.dump(file_hashes, f)

# Funktion zur Ausführung eines Befehls
def execute_command():
    subprocess.run(["cd", "webapp"], shell=True)
    subprocess.run(["yarn", "install"], shell=True)
    subprocess.run(["yarn", "build"], shell=True)


# Pfad zum Ordner und zur Hash-Datei
directory = './webapp'
hash_file = "file_hashes.pkl"

# Überprüfung der Dateien
check_files(directory, hash_file)

```

kanns auf meinem System nur nich direkt testen, ``yarn`` wird nicht gefunden. die entstehende Hashtable is ~9kB groß.

**Reply by schlimmchen** [reply]
Ich begreife nicht, wie das skript hierbei hilft. Wenn jemand das Repo frischt klont, dann merkt er/sie wie genau, dass die webapp neu gebaut werden muss?

Ein pre-skript wäre wohl nett, um eine klare, gut verständliche fehlermeldung auszugeben, wenn die dateien fehlen. Sonst versteht das wiederum jemand nicht, dass es um die fehlende web app geht.

Allerdings ist so ein Ansatz natürlich interessant für Leute die häufiger kompilieren, denn man vergisst immer mal wieder die webapp zu aktualisieren bevor man die firmware baut.

**Reply by spcqike** [reply]
>  Wenn jemand das Repo frischt klont, dann merkt er/sie wie genau, dass die webapp neu gebaut werden muss?

ich hätte gedacht, wenn jemand das Repo frisch klont, ist sowohl die ``webapp`` als auch die ``webapp_dist`` up to date, oder nicht?

das Skript erstellt von allen Datein im ``webapp`` Ordner einen Hash und speichert es in der ``file_hashes``. sind alle datei-hashes gleich der in der ``file_hashes``, hat sich im ``webapp`` nichts geändert und es muss nicht neu gebaut werden. ist ein hash unterschiedlich, führt es ``execute_command():`` aus. die Funktion muss wie gesagt getestet und ggf. angepasst werden.

vielleicht hab ich aber auch einfach das Grundproblem nicht oder falsch verstanden 😄

**Comment by Manos1966** [comment]
@schlimmchen liegt es vielleicht weil wir es so im Wiki erklaeren?

https://github.com/helgeerbe/OpenDTU-OnBattery/wiki/Compile-your-own-Firmware-with-Visual-Studio-Code

**Reply by schlimmchen** [reply]
Ist es gemein, wenn ich der Meinung bin, dass die verlinkte Wikiseite komplett redundant bzw. obsolet ist?

In der upstream-Doku enthaltene Infos sollten wir nicht duplizieren, höchstens verlinken.

**Reply by Manos1966** [reply]
> In der upstream-Doku enthaltene Infos sollten wir nicht duplizieren, höchstens verlinken.

Wie "dupliziert"? Es ist der Link zur Upstream-Doku
...und irgendwo mussten wir den Unterschied zur Upstream schreiben:
> instead of tbnobody's original OpenDTU https://github.com/tbnobody/OpenDTU.git
you need to clone
the OpenDTUonBattery directory https://github.com/helgeerbe/OpenDTU-OnBattery.git

Die Frage die sich stellt ist:
1. Sollten wir Hinweise zur "Selbst-Compilierung" im Wiki haben?
2. Falls ja, wie viele Optionen?
  2.1 [Compile with Visual Studio Code](https://www.opendtu.solar/firmware/compile_vscode/)
  2.2 [Compile on the command line with PlatformIO Core](https://www.opendtu.solar/firmware/compile_cli/)
  2.3 [Compile WebApp](https://www.opendtu.solar/firmware/compile_webapp/)

**Reply by Manos1966** [reply]
Ergänzung, um euch die Suche zu ersparen: 

Zur Zeit haben wir ganz Oben unter "First Time Installation" ausschließend den Link zum [Webflasher](https://solar.metacontrol.eu/opendtu-onbattery-webflasher/)

Ganz Unten unter "Sonstiges" haben wir die Erklärungen bezüglich Selbst Kompilierung:
![Compilierung](https://github.com/user-attachments/assets/2114ff8e-fd0c-4d18-9cc4-dac86b6706a8)
